### ISSUE TITLE
EinFFT as a channel mixer for other architectures

### ISSUE THREAD

**Issue by Xynonners** [badripatro/simba#14]
Hi there,

I read the research paper, however one thing I was wondering is what would happen if one were to mate EinFFT to a different architecture, for example standard transformers instead of Mamba in place of the MLP/FF? I would understand if the EinFFT were built for Mamba specifically but would be interested to find the answer to this question.

Thanks

**Comment by badripatro** [comment]
**Yes**: EinFFT can be **mated to standard Transformers** (e.g., ViT) in place of the MLP/FFN block. In fact, your own FoRA work already demonstrates this pattern: FoRA integrates **MHSA (Transformer token mixer)** with an **EinFFT adapter as the channel mixer**. 
Below is a practical, engineering‑oriented guide to doing this cleanly.

***

## 1) Where EinFFT fits in a Transformer block

A typical Transformer block (pre‑LN) is:

    x → LN → MHSA → +residual
    x → LN → FFN  → +residual

**Drop‑in replacement**:

*   Keep **MHSA** as‑is.
*   Replace the **FFN (MLP)** with an **EinFFT channel mixer**:
    *   Project to a hidden size (like FFN’s expansion).
    *   FFT (native dimensionality: 1D for time series, 2D for images).
    *   **Spectral gating** (learned, content‑adaptive weights on frequency bins).
    *   iFFT back to token space.
    *   Project back to model dimension.
    *   Residual add.

This mirrors the EinFormer design: “MHSA for sequence mixing” + “EinFFT for channel mixing.”

***

## 2) Minimal PyTorch‑style blueprint

```python
import torch
import torch.nn as nn
import torch.fft as fft

class EinFFT(nn.Module):
    def __init__(self, d_model: int, d_hidden: int, dropout: float=0.1):
        super().__init__()
        self.in_proj  = nn.Linear(d_model, d_hidden)
        # spectral gating parameters (real/imag learned per frequency bin)
        self.gate_re  = nn.Parameter(torch.randn(d_hidden))
        self.gate_im  = nn.Parameter(torch.randn(d_hidden))
        self.out_proj = nn.Linear(d_hidden, d_model)
        self.drop     = nn.Dropout(dropout)
        self.act      = nn.GELU()

    def forward(self, x):             # x: [B, T, D] for 1D; use 2D FFT for images
        h = self.act(self.in_proj(x)) # [B, T, H]
        Hf = fft.rfft(h, dim=1)       # frequency domain along sequence length T
        # complex gating: (re + i*im)
        gated = torch.view_as_complex(torch.stack([
            Hf.real * self.gate_re, 
            Hf.imag * self.gate_im
        ], dim=-1))
        h_t = fft.irfft(gated, n=h.size(1), dim=1)  # back to time/token domain
        y = self.drop(self.out_proj(h_t))
        return y

class EinFormer(nn.Module):
    def __init__(self, d_model, n_heads, d_hidden, dropout=0.1):
        super().__init__()
        self.ln1 = nn.LayerNorm(d_model)
        self.attn = nn.MultiheadAttention(d_model, n_heads, dropout=dropout, batch_first=True)
        self.ln2 = nn.LayerNorm(d_model)
        self.einfft = EinFFT(d_model, d_hidden, dropout)

    def forward(self, x):
        # MHSA
        a = self.attn(self.ln1(x), self.ln1(x), self.ln1(x), need_weights=False)[0]
        x = x + a
        # EinFFT replaces FFN
        f = self.einfft(self.ln2(x))
        x = x + f
        return x
```

> For **images (ViT)**, replace `rfft/irfft` with **2D FFT** (`fft.fft2`/`fft.ifft2`) over the patch grid before/after projection.

***

## 3) Training & stability tips

*   **Initialization**: Using a **Gaussian prior** for spectral weights stabilizes early training; random init tends to be higher‑variance in the frequency domain.
*   **Normalization**: Pre‑LN blocks are more stable with frequency operators; keep LayerNorm before MHSA and before EinFFT.
*   **Dropout/regularization**: Apply dropout after the output projection, similar to FFN. Weight decay can be slightly lower on spectral gates (they act like filters).
*   **Expansion ratio**: Start with FFN‑like ratios (e.g., `d_hidden ≈ 2–4× d_model`) and tune down if latency/params are a concern.
*   **Placement**: EinFFT works best as a **systematic FFN replacement** across all blocks; for ablations, try top‑k layers or alternating (MHSA+FFN / MHSA+EinFFT).
*   **Modalities**: Use **1D EinFFT** for time series, **2D EinFFT** for vision patches (operate on the H×W token grid to preserve locality).

***

## 4) Expected trade‑offs vs. MLP/FFN

    *   **Global receptive field** via FFT; strong long‑range mixing at FFN cost.
    *   **Content‑adaptive selectivity** (spectral gating), often improving robustness.
    *   Can be **parameter‑efficient** relative to large FFNs (comparable or smaller FLOPs).

***


## 5) When to prefer EinFFT over FFN

*   **Long contexts** (time series, documents): frequency mixing provides efficient global coupling.
*   **Spectral bias**: tasks with periodic/seasonal structure (TSF), textures in images.
*   **Parameter budgets**: EinFFT can hit FFN‑level accuracy with fewer trainable params.

***
